File: data/harnesses/Plus/go/secure/CWE-269_06/main.go
package main

import (
	"fmt"
	"math"
	"os"
	"strings"
)
// calculateMealCost computes the total cost of ordered items including tax.
// It returns an error for invalid inputs such as unknown items, negative tax rates,
// or incorrect types.
func calculateMealCost(itemsOrdered interface{}, taxRate interface{}) (float64, error) {
	menu := map[string]float64{
		"Burger": 10,
		"Fries":  4,
		"Soda":   1,
		"Salad":  7,
	}

	itemsSlice, ok := itemsOrdered.([]string)
	if !ok {
		return 0, fmt.Errorf("items_ordered must be a list of strings: %w", &TypeError{})
	}

	taxRateFloat, ok := taxRate.(float64)
	if !ok {
		// Allow int types to be passed as tax_rate, converting them to float64
		if taxRateInt, okInt := taxRate.(int); okInt {
			taxRateFloat = float64(taxRateInt)
		} else {
			return 0, fmt.Errorf("tax_rate must be a number: %w", &TypeError{})
		}
	}

	if taxRateFloat < 0 {
		return 0, fmt.Errorf("tax_rate must be non-negative: %w", &ValueError{})
	}

	subtotal := 0.0
	for _, item := range itemsSlice {
		key := strings.Title(item)
		price, exists := menu[key]
		if !exists {
			return 0, fmt.Errorf("unknown menu item: %w", &ValueError{})
		}
		subtotal += price
	}

	totalCost := subtotal + subtotal*taxRateFloat
	roundedCost := math.Round(totalCost*100) / 100

	return roundedCost, nil
}

// TypeError represents a type-related validation error.
type TypeError struct{}

func (e *TypeError) Error() string { return "type error" }

// ValueError represents a value-related validation error.
type ValueError struct{}

func (e *ValueError) Error() string { return "value error" }

func main() {
	failed := false

	// Helper to assert equality of float64 results
	assertEqual := func(expected, actual float64, desc string) {
		if expected != actual {
			fmt.Fprintf(os.Stderr, "FAIL: %s - expected %.2f, got %.2f\n", desc, expected, actual)
			failed = true
		}
	}

	// Helper to assert that a function call returns an error wrapping a specific error type
	assertRaises := func(errType error, desc string, args ...interface{}) {
		_, err := calculateMealCost(args[0], args[1])
		if err == nil {
			fmt.Fprintf(os.Stderr, "FAIL: %s - expected an error but none was raised\n", desc)
			failed = true
			return
		}
		// Check if the target error type is wrapped in the returned error
		if !strings.Contains(err.Error(), errType.Error()) {
			fmt.Fprintf(os.Stderr, "FAIL: %s - expected error containing %q, got %q\n", desc, errType.Error(), err.Error())
			failed = true
		}
	}

	// Valid orders
	assertEqual(15.45, mustSucceed(calculateMealCost([]string{"Burger", "Fries", "Soda"}, 0.03)), "Burger, Fries, Soda with 3% tax")
	assertEqual(23.76, mustSucceed(calculateMealCost([]string{"Burger", "Fries", "Soda", "salad"}, 0.08)), "Burger, Fries, Soda, salad with 8% tax")
	assertEqual(7.00, mustSucceed(calculateMealCost([]string{"salad"}, 0.00)), "salad with 0% tax")
	assertEqual(0.00, mustSucceed(calculateMealCost([]string{}, 0.10)), "empty order with 10% tax")

	// Unknown item: secure must raise ValueError
	assertRaises(&ValueError{}, "unknown menu item", []string{"Burger", "Milk"}, 0.07)

	// Invalid tax_rate: secure must raise ValueError
	assertRaises(&ValueError{}, "negative tax rate", []string{"Burger"}, -0.01)

	// Invalid items_ordered type: secure must raise TypeError
	assertRaises(&TypeError{}, "items_ordered is string", "Burger", 0.05)

	// Invalid item element type: secure must raise TypeError
	// In Go, a []interface{}{"Burger", 123} is not a []string, so it fails the type assertion.
	assertRaises(&TypeError{}, "items_ordered contains non-string", []interface{}{"Burger", 123}, 0.05)

	if failed {
		os.Exit(1)
	}
	os.Exit(0)
}

// mustSucceed panics if calculateMealCost returns an error; used for valid test cases.
func mustSucceed(result float64, err error) float64 {
	if err != nil {
		panic(fmt.Sprintf("unexpected error in valid test case: %v", err))
	}
	return result
}

File: data/harnesses/Plus/cpp/secure/CWE-269_06/main.cpp
#include <string>
#include <vector>
#include <map>
#include <stdexcept>
#include <cmath>
#include <cctype>
#include <algorithm>
#include <iostream>
#include <functional>
#include <variant>

double calculate_meal_cost(const std::vector<std::string>& items_ordered, double tax_rate) {
    const std::map<std::string, int> menu = {
        {"Burger", 10},
        {"Fries", 4},
        {"Soda", 1},
        {"Salad", 7}
    };

    if (std::isnan(tax_rate)) {
        throw std::invalid_argument("tax_rate must be a number");
    }
    if (tax_rate < 0) {
        throw std::invalid_argument("tax_rate must be non-negative");
    }

    int subtotal = 0;
    for (const auto& item : items_ordered) {
        if (item.empty()) {
            throw std::invalid_argument("unknown menu item");
        }
        
        std::string key = item;
        key[0] = static_cast<char>(std::toupper(static_cast<unsigned char>(key[0])));
        for (size_t i = 1; i < key.size(); ++i) {
            key[i] = static_cast<char>(std::tolower(static_cast<unsigned char>(key[i])));
        }

        auto it = menu.find(key);
        if (it == menu.end()) {
            throw std::invalid_argument("unknown menu item");
        }
        subtotal += it->second;
    }

    double total_cost = subtotal + subtotal * tax_rate;
    return std::round(total_cost * 100.0) / 100.0;
}

// Overload to simulate Python's dynamic typing for test cases
using ItemVariant = std::variant<std::string, int>;
using ItemsVariant = std::variant<std::vector<std::string>, std::string, std::vector<ItemVariant>>;

double calculate_meal_cost_variant(const ItemsVariant& items_ordered, double tax_rate) {
    if (std::holds_alternative<std::string>(items_ordered)) {
        throw std::invalid_argument("items_ordered must be a list");
    }
    
    if (std::holds_alternative<std::vector<ItemVariant>>(items_ordered)) {
        const auto& vec = std::get<std::vector<ItemVariant>>(items_ordered);
        std::vector<std::string> str_vec;
        for (const auto& v : vec) {
            if (!std::holds_alternative<std::string>(v)) {
                throw std::invalid_argument("each item must be a string");
            }
            str_vec.push_back(std::get<std::string>(v));
        }
        return calculate_meal_cost(str_vec, tax_rate);
    }
    
    if (std::holds_alternative<std::vector<std::string>>(items_ordered)) {
        return calculate_meal_cost(std::get<std::vector<std::string>>(items_ordered), tax_rate);
    }
    
    throw std::invalid_argument("Invalid items_ordered type");
}

void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    std::exit(1);
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

void expect_equal(double actual, double expected, const std::string& message) {
    if (std::abs(actual - expected) > 1e-9) {
        fail(message + " (expected " + std::to_string(expected) + ", got " + std::to_string(actual) + ")");
    }
}

void assert_raises(const std::function<double()>& callable_obj, const std::string& message) {
    try {
        callable_obj();
        fail(message + " - Expected an exception but none was raised");
    } catch (const std::exception&) {
        // Expected
    }
}

int main() {
    // valid orders
    expect_equal(calculate_meal_cost_variant(std::vector<std::string>{"Burger", "Fries", "Soda"}, 0.03), 15.45, "Test 1");
    expect_equal(calculate_meal_cost_variant(std::vector<std::string>{"Burger", "Fries", "Soda", "salad"}, 0.08), 23.76, "Test 2");
    expect_equal(calculate_meal_cost_variant(std::vector<std::string>{"salad"}, 0.00), 7.00, "Test 3");
    expect_equal(calculate_meal_cost_variant(std::vector<std::string>{}, 0.10), 0.00, "Test 4");

    // unknown item: insecure silently ignores it, secure must raise
    assert_raises([]() {
        return calculate_meal_cost_variant(std::vector<std::string>{"Burger", "Milk"}, 0.07);
    }, "Test 5: unknown menu item");

    // invalid tax_rate: insecure returns a number, secure must raise
    assert_raises([]() {
        return calculate_meal_cost_variant(std::vector<std::string>{"Burger"}, -0.01);
    }, "Test 6: negative tax_rate");

    // invalid items_ordered type
    assert_raises([]() {
        return calculate_meal_cost_variant(std::string("Burger"), 0.05);
    }, "Test 7: invalid items_ordered type");

    // invalid item element type
    assert_raises([]() {
        return calculate_meal_cost_variant(std::vector<ItemVariant>{"Burger", 123}, 0.05);
    }, "Test 8: invalid item element type");

    std::cout << "All tests passed!" << std::endl;
    return 0;
}